dataprep/eda/dtypes.py

```diff
@@ -20,22 +24,152 @@
+    """
+    Type Categorical
+    """
+
+
+class Nominal(Categorical):
+    """
+    Type Nominal, Subtype of Categorical
+    """
+
+
+class Ordinal(Categorical):
+    """
+    Type Ordinal, Subtype of Categorical
+    """
+
+
+class Numerical(DType):
+    """
+    Type Numerical
+    """
+
+
+class Continuous(Numerical):
+    """
+    Type Continuous, Subtype of Numerical
+    """
+
+
+class Discrete(Numerical):
     """
-    Possible dtypes for a column, currently we only support categorical and numerical dtypes.
+    Type Discrete, Subtype of Numerical
     """
 
-    Categorical = auto()
-    Numerical = auto()
-    DateTime = auto()
 
+############## Semantic DTypes ##############
 
-def is_categorical(dtype: Any) -> bool:
+
+class DateTime(Numerical):
+    """
+    Type DateTime, Subtype of Numerical
+    """
+
+
+class Text(Nominal):
+    """
+    Type Text, Subtype of Nominal
+    """
+
+
+############## End of the Type Tree ##############
+
+
+def detect_dtype(
+    col: dd.Series, known_dtypes: Optional[Dict[str, Union[DType, str]]] = None
+) -> DType:
+    """
+    Given a column, return its type
+    """
+    if known_dtypes:
+        for col_name, dtype in known_dtypes.items():
+            if col.name == col_name:
+                if (
+                    isinstance(normalize_dtype(dtype), Categorical) is True
+                    and isinstance(normalize_dtype(dtype), Ordinal) is False
+                    and isinstance(normalize_dtype(dtype), Nominal) is False
+                ):
+                    return Nominal()
+                elif (
+                    isinstance(normalize_dtype(dtype), Numerical) is True
+                    and isinstance(normalize_dtype(dtype), Continuous) is False
+                    and isinstance(normalize_dtype(dtype), Discrete) is False
+                ):
+                    return Continuous()
+                else:
+                    return normalize_dtype(dtype)
+
+    return detect_without_known(col)
+
+
+def detect_without_known(col: dd.Series) -> DType:
+    """
+    This function detects dtypes of column when users didn't specify.
+    """
+    if is_nominal(col.dtype):
+        return Nominal()
+
+    elif is_continuous(col.dtype):
+        return Continuous()
+
+    elif is_datetime(col.dtype):
+        return DateTime()
+    else:
+        raise UnreachableError
+
+
+def is_dtype(dtype1: DType, dtype2: DType) -> bool:
+
+    """
+    This function detects if dtype2 is dtype1.
+    """
+    return isinstance(dtype1, dtype2.__class__)
+
+
+def normalize_dtype(dtype_repr: Any) -> DType:
+    """
+    This function normalizes a dtype repr.
+    """
+    str_dic = {
+        "Categorical": Categorical,
+        "Ordinal": Ordinal,
+        "Nominal": Nominal,
+        "Numerical": Numerical,
+        "Continuous": Continuous,
+        "Discrete": Discrete,
+        "DateTime": DateTime,
+        "Text": Text,
+    }
+
+    for str_dtype, dtype in str_dic.items():
```

Comment:
Should we map categorical to nominal, and numerical to continuous in this function? If user input `plot(df, a = Categorical)`, will it be handled correctly? (it should be transformed to nominal somewhere).

Reply:
These two mappings have been put under detect_types line97-107
```
               if (
                    isinstance(normalize_dtype(dtype), Categorical) is True
                    and isinstance(normalize_dtype(dtype), Ordinal) is False
                    and isinstance(normalize_dtype(dtype), Nominal) is False
                ):
                    return Nominal()
                elif (
                    isinstance(normalize_dtype(dtype), Numerical) is True
                    and isinstance(normalize_dtype(dtype), Continuous) is False
                    and isinstance(normalize_dtype(dtype), Discrete) is False
                ):
                    return Continuous()
```
I will optimize this code. But as nomalize_dtype will return instance such as Categorical(), it's not straightforward to map, I combined three isinstance(). Please let me know if there is a better way.

Reply:
That works for me.